pkg/microservice/aslan/core/templatestore/service/chart.go

```diff
@@ -0,0 +1,246 @@
+		CodeHostID: args.CodehostID,
+		Sha1:       sha1,
+	})
+}
+
+func UpdateChartTemplate(name string, args *fs.DownloadFromSourceArgs, logger *zap.SugaredLogger) error {
+	chart, err := mongodb.NewChartColl().Get(name)
+	if err != nil {
+		logger.Errorf("Failed to get chart template %s.", name)
+		return err
+	}
+
+	sha1, err := processChartFromSource(name, args, logger)
+	if err != nil {
+		logger.Errorf("Failed to update chart %s, err: %s", name, err)
+		return err
+	}
+
+	if chart.Sha1 == sha1 {
+		logger.Debug("Chart %s has no changes, skip updating.", name)
+		return nil
+	}
+
+	return mongodb.NewChartColl().Update(&models.Chart{
+		Name:       name,
+		Owner:      args.Owner,
+		Repo:       args.Repo,
+		Path:       args.Path,
+		Branch:     args.Branch,
+		CodeHostID: args.CodehostID,
+		Sha1:       sha1,
+	})
+}
+
+func RemoveChartTemplate(name string, logger *zap.SugaredLogger) error {
+	err := mongodb.NewChartColl().Delete(name)
+	if err != nil {
+		logger.Errorf("Failed to delete chart template %s.", name)
+		return err
+	}
+
+	if err = fs.DeleteArchivedFileFromS3(name, configbase.ObjectStorageChartTemplatePath(name), logger); err != nil {
+		logger.Warnf("Failed to delete file %s, err: %s", name, err)
+	}
+
+	return nil
+}
+
+func processChartFromSource(name string, args *fs.DownloadFromSourceArgs, logger *zap.SugaredLogger) (string, error) {
+	tree, err := fs.DownloadFilesFromSource(args, func(a afero.Fs) (string, error) {
+		return "", nil
+	})
+
+	if err != nil {
+		logger.Errorf("Failed to download files with option %+v, err: %s", args, err)
+		return "", err
+	}
+
+	var sha1 string
+	if name == "" {
+		name = filepath.Base(args.Path)
+	}
+
+	var wg wait.Group
+	wg.Start(func() {
+		logger.Debug("Start to save and upload chart")
+		localBase := configbase.LocalChartTemplatePath(name)
+		s3Base := configbase.ObjectStorageChartTemplatePath(name)
+
+		err1 := fs.SaveAndUploadFiles(tree, name, localBase, s3Base, logger)
+		if err1 != nil {
+			logger.Errorf("Failed to save files to disk, err: %s", err1)
+			err = err1
+			return
+		}
+
+		logger.Debug("Finish to save and upload chart")
+	})
+
+	wg.Start(func() {
+		logger.Debug("Start to calculate sha1 for chart")
+		tmpDir, err1 := os.MkdirTemp("", "")
+		if err1 != nil {
+			logger.Errorf("Failed to create temp dir, err: %s", err1)
+			err = err1
+			return
+		}
+		defer func() {
+			_ = os.RemoveAll(tmpDir)
+		}()
+
+		fileName := fmt.Sprintf("%s.tar.gz", filepath.Base(args.Path))
+		tarball := filepath.Join(tmpDir, fileName)
+		if err1 = fsutil.Tar(tree, tarball); err1 != nil {
+			logger.Errorf("Failed to archive files to %s, err: %s", tarball, err1)
+			err = err1
+			return
+		}
+
+		sha1, err1 = fsutil.Sha1(os.DirFS(tmpDir), fileName)
+		if err1 != nil {
+			logger.Errorf("Failed to calculate sha1 for file %s, err: %s", tarball, err1)
+			err = err1
+			return
+		}
+
+		logger.Debug("Finish to calculate sha1 for chart")
+	})
+
+	wg.Wait()
+
+	if err != nil {
```

Comment:
You can delete this and put it in defer at the beginning

Reply:
no difference